.github/workflows/version-bump.yml

```diff
@@ -0,0 +1,50 @@
+name: Version bump
+
+on:
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'Version (major.minor.patch)'
+        required: true
+      abi-version:
+        description: 'ABI version (major.minor.patch)'
+        required: true
+
+env:
+  VERSION: ${{ github.event.inputs.version }}
+  ABI_VERSION: ${{ github.event.inputs.abi-version }}
+  GITHUB_TOKEN: ${{ github.token }}
+
+jobs:
+  bump-version:
+    name: "Bump version"
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Check out Criterion
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+
+      - name: Configure git
+        run: |
+          git config user.name "github-actions"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+
+      - name: Create branch for version bump
+        run: git switch -c "release/bump-v${VERSION}"
+
+      - name: Bump version
+        run: |
+          version_filter="(version\s*:\s*')[0-9a-z.+-]+'"
+          sed -E -i "s|${version_filter}|\1${VERSION}'|g" meson.build
```

Comment:
I think I'd rather have this be fully inferred from git state. There's code in meson.build to run git describe to get the proper version for any given commit, so I think we should just put an obviously invalid fallback, and provide a way to freeze the version string for source tarballs without git metadata (e.g. reading the content of a .version file in the archive for instance). In particular, I would really like to avoid robot commits.

Reply:
Okay, let's create a version file to freeze the version in the tarball.

Just to make sure I understand: do you prefer not to use automated git operations at all, or do you just want to avoid bot commits that do their own code modification?

For example, would that be acceptable for a bot to create an annotated tag, merge the master branch back to bleeding when the release is done, et cetera, or should we rather do all of those manually?

Reply:
I'm not against having some git operations done by robots, but there aren't that many left if you remove commits and tags from the equation. Robot commits tend to pollute the history so I'd rather avoid them whenever possible, and I think tags shouldn't be done by robots either, because you can't do some things when tagging as a robot.

For once, you can't sign tags unless you breach trust and allow some external system access to your private key, or you have to generate a key pair for a robot, at which point you essentially tell others to trust a robot whose key management is not transparent rather than trusting a human maintainer. Though, now that I'm double checking, I actually forgot to sign the release candidate tag, but I'd rather have something that checks the validity of the tag rather than something that creates it for me, I think.

It also feels backwards to automatically create tags; what's the trigger for that operation? I'd have assumed that what would trigger all of the automation to build, package, and publish a release would have been a human pushing a tag.

Also, that's not to say that we can't have automation locally; I'm fine with a script to prepare a tag with the right message, perform any --no-ff merge, and whatever else is needed, since at the end of the day the operations would be authored by whoever runs the script.

Reply:
> create a version file to freeze the version in the tarball.

It seems Meson won't allow uncommitted changes in the tarball, so we need to commit the version-freeze (at least as a local commit):
https://github.com/mesonbuild/meson/issues/4277

> It also feels backwards to automatically create tags; what's the trigger for that operation?

The trigger would have been a manual action, such as adding a label, or pushing a "release button".
My intention with this was to document the release process as code, which can be executed in CI.

I see your point, let's change the workflow:
- remove the release bump workflow
- figure out what the "release draft workflow" trigger should be

I think a tag-push trigger would be too late for us because the source tarball or binary generation may fail due to a real bug in Criterion, which then we want to fix before the release. I believe tags should not be removed/re-created, so I think the best would be if we triggered the workflow with something else, for example:

trigger by
- pushing a commit to the master branch with a specific instruction in it, such as "@github-actions draft release v2.4.0"
- sending an instruction as a comment in a GitHub issue
- always creating a pull request manually with a "release" label and merging it
- starting the job manually on GitHub, where you can input the version number (since we don't have a tag yet)

Reply:
I also agree that tags should not be removed or recreated, but given that we have a release candidate to fix any issues before an actual release, I think we should make the most out of it. How about this:

* Every time new commits gets pushed to master, the new branch tip gets a release (as in, published on the github release page), alongside any built packages. That release is always marked as a prerelease, and its version is generated from `git describe --long`.
* If a new tag not containing `-rc` is pushed, a new release (which this time isn't marked as a prerelease) gets published.

If we do this, then it means that the release workflow becomes:

1. merge bleeding into master, tag it with `vX.Y.Z-rc`, push both master and the tag.
2. the release action runs, builds a release source tarball & binary build for Criterion on CentOS, then creates a new prerelease on github based of the output of `git describe --long`, which will look like `vX.Y.Z-rc-0-gdeadbeef`, and attached the aforementioned artifacts to that release.
3. someone makes a pull request to fix a bug in the vX.Y.Z release candidate, we review & merge it to master; let's say for this argument that it contains 3 new commits and the last one is `feedface`
4. the release action runs because master has new commits, and goes on to make a new prerelease named `vX.Y.Z-rc-3-gfeedface`.
5. after some time, when the prerelease feels ready for release, we tag vX.Y.Z on the tip of master, which causes an action to run and create the final release named vX.Y.Z.

Some additional would-be-nice-to-have on top of that:
* automatically fast-forward bleeding to be on the tip of master. If that's not possible, warn the maintainers to manually intervene if necessary (not being able to fast-forward isn't necessarily bad, it could be that development continued on bleeding and we cherry-picked a few commits from bleeding to master)
* check that the release tag is signed; warn maintainers if not (don't think we can do much more than warn given that changing tags is not allowed, although we could make an exception for the automation and automatically remove it precisely because we deem unsigned release tags to be invalid)

What do you think?

Reply:
Sounds good to me. I'll try all this in my fork in practice and share the results.

Reply:
My plan is to try all the above over the weekend.

Reply:
I've updated the description according to my changes. I'll test the workflow properly and attach a few screenshots tomorrow.

Reply:
I've updated the description according to my changes. Tomorrow, I'll test the workflow properly and attach a few screenshots.

Meson supports version files, but that file must always be present and the feature is only available in new Meson versions, so I've decided to leave the version string intact. A manual version bump is needed before each release, but this is not a hard requirement, because the tarball is created using `meson rewrite kwargs set project / version`, which replaces the version number in `meson.build`.